docs/internal_abi.md

```diff
@@ -0,0 +1,323 @@
+
+### Example: `Float64 | T* | U*`
+
+The payload size of `Float64` is 64 bits and it requires one discriminator state.
+The payload size of `T*` and `U*` is each either 32, 48, or 128 bits, depending on the target architecture and they require one discriminator state between them.
+The lowering for this union therefore depends on the target.
+
+On 32-bit systems, the largest type is `Float64` and the remaining types all fit easily in the significand bits.
+There are no discriminator bits required but the 'as-if' rules triggers as if there were a single bit for determining which end of the space should be used and the remaining values are stored at the end of the space that is strongly aligned: bits 0-31.
+
+On 64-bit systems, the same rules apply, packing the used bits of the pointer into bits 0-50.
+
+On CHERI systems, the pointer types are larger than the `Float64` and so the payload is a 128-bit value, distinguished by the tag bit.
+
+### Example: `T* | None`
+
+In this example, `None` is a singleton type with no fields, used as a placeholder indicating a not-present value.
+The `T*` requires one pointer's worth of space for the payload (as above, this varies depending on the target) and one discriminator state.
+The `None` value requires one discriminator state and no payload.
+
+The largest value is a pointer and so on CHERI systems the tag bit is used as the discriminator, on other architectures the low bit of the pointer is used (because the pointer is the largest type and the discriminator fits in a three-bit value).
+The discriminator values are assigned starting from the singletons, so zero is used for `None` and one for `T*`.
+
+A pattern match expression on this type is compiled into a compare-and-branch-on-{non}-zero, which is a single instruction on most architectures.
+Accessing the `T*` on a CHERI system requires no further operations.
+On integer-pointer systems, it requires subtracting one.
+On any mainstream architecture other than RISC-V, this subtraction can be folded into the immediate offset on a load instruction and so requires no additional instructions.
+
+If the `None` value is used (rather than just being pattern matched on and ignored) then the compiler must be able to materialise the address (or capability) to the singleton.
+This is no worse than if we stored the address inline; however, because in that case the match would require a comparison against the address and so the address would need to be loaded before the pattern match, rather than conditionally in one path.
+
+Objects
+-------
+
+Verona objects contain a header with metadata describing the object, followed by user-defined fields.
+
+### Object headers
+
+Object headers are two pointer-size words.
+These are described in detail in [src/rt/object/object.h](../src/rt/object/object.h).
+
+The first word contains the region metadata.
```

Comment:
The thing it points to is 8-byte aligned even on 32-bit systems so that we can still steal the three low bits.  It should be possible to use a pointer-sized value here, but I'm not sure.